crates/matrix-sdk-ui/src/timeline/read_receipts.rs

```diff
@@ -465,25 +466,46 @@ impl TimelineInnerState {
         let private_read_receipt =
             self.user_receipt(user_id, ReceiptType::ReadPrivate, room_data_provider).await;
 
-        // If we only have one, return it.
-        let Some(pub_receipt) = &public_read_receipt else {
-            return private_read_receipt;
-        };
-        let Some(priv_receipt) = &private_read_receipt else {
-            return public_read_receipt;
-        };
-
-        match self.compare_receipts(pub_receipt, priv_receipt) {
-            Some(Ordering::Greater) => public_read_receipt,
-            Some(Ordering::Less) => private_read_receipt,
-            _ => {
-                // As a fallback, let's assume that a private read receipt should be more recent
-                // than a public read receipt, otherwise there's no point in the private read
-                // receipt.
-                private_read_receipt
-            }
+        match self.pub_or_priv_receipt(public_read_receipt.as_ref(), private_read_receipt.as_ref())
+        {
+            Ordering::Greater => public_read_receipt,
+            Ordering::Less => private_read_receipt,
+            _ => unreachable!(),
         }
     }
+
+    pub(super) fn latest_user_read_receipt_timeline_event_id(
+        &self,
+        user_id: &UserId,
+    ) -> Option<OwnedEventId> {
+        // We only need to use the local map. Since receipts for known events are
+        // already loaded from the store.
+        let public_read_receipt = self
+            .read_receipts
+            .users_read_receipts
+            .get(user_id)
+            .and_then(|user_map| user_map.get(&ReceiptType::Read));
+        let private_read_receipt = self
+            .read_receipts
+            .users_read_receipts
+            .get(user_id)
+            .and_then(|user_map| user_map.get(&ReceiptType::ReadPrivate));
+
+        let (latest_receipt_id, _) =
+            match self.pub_or_priv_receipt(public_read_receipt, private_read_receipt) {
+                Ordering::Greater => public_read_receipt?,
+                Ordering::Less => private_read_receipt?,
+                _ => unreachable!(),
+            };
+
+        // Find the corresponding visible event.
+        self.all_events
+            .iter()
+            .rev()
+            .skip_while(|ev| ev.event_id != *latest_receipt_id)
+            .find(|ev| ev.visible)
+            .map(|ev| ev.event_id.clone())
```

Comment:
I think this can be summarized to:

```suggestion
            .find_map(|ev| (ev.event_id == *latest_receipt_id && ev.visible).then(|| ev.event_id.clone()))
```

Also, is filtering by `visible`-only events something new and not documented in the functions' doc comments?

Reply:
This is not the same thing:
1. We search for the real event for the read receipt
2. We get the first visible event starting from it (which might be the same event)

Your code would only work if the real event for the read receipt is visible.

It seems like I forgot to document this method.